### PR TITLE
1.11.2 fork avatair

### PR DESCRIPTION
**Pull request by ghost**
This forked version contains mainly a bug fix and a feature. Changes are submitted to 1.11.2, because no 1.12.x version is available.

Bugfix:
Fixing replanting bug like https://github.com/TeamTwilight/twilightforest/issues/79 . Central problematic method is  BlockTFPlant.canBlockStay which doesn't receive any information about the meta of the block to place. Solved this issue by having four different plant block types, see EnumTFPlantType . 

Feature:
A 2x2 trunk version of canopy tree was readded, as formely introduced in MC 1.2.5 and then removed. See TFGenHugeCanopyTree. Canopy tree saplings can be planted as 2x2 square to grow this kind of tree as well. This feature requires TFBiomeDecorator.decorate to be made recursively callable, because the size of the tree is big enough to invoke additional chunk generations during decoration.

### CODE REVIEW

**Comment by ChloeDawn** [comment]
> because no 1.12.x version is available.

1.12.x is literally the main repository branch...

**Comment by ghost** [comment]
Excuse my English. I mean I don't use 1.12.x and currently play and develop on 1.11.2 . So 1.12.x is not available for me ;)

**Comment by williewillus** [comment]
Please target your PR to 1.12

File: src/main/resources/assets/twilightforest/lang/en_us.lang [hunk 1]
```diff
@@ -27,15 +27,15 @@ tile.TFBossSpawner.2.name=Hydra Spawner
...
+tile.TFPlantCave.0.name=Moss Patch
+tile.TFPlantStandard.1.name=Mayapple
+tile.TFPlantStandard.2.name=Clover Patch [WIP]
+tile.TFPlantStandard.3.name=Fiddlehead
+tile.TFPlantCave.4.name=Mushgloom
+tile.TFPlantDark.5.name=Forest Grass
+tile.TFPlantDark.6.name=Dried Bush
+tile.TFPlantHanging.7.name=Torchberry Plant
+tile.TFPlantHanging.8.name=Root Strands
 tile.TFRoots.0.name=Roots
```
Comment (williewillus) [inline src/main/resources/assets/twilightforest/lang/en_us.lang:39]: Change all the other lang files

File: src/main/java/twilightforest/biomes/RegistryBiomeEvent.java [hunk 2]
```diff
@@ -44,8 +44,8 @@ public static void onRegisterBiomes(Register<Biome> event) {
 								.setWaterColor(0x005522)
 								.setTemperature(0.7F)
 								.setRainfall(0.8F)
-								.setBaseHeight(0.2F)
-								.setHeightVariation(0.2F)
+								.setBaseHeight(0.15F)
+								.setHeightVariation(0.4F)
```
Comment (williewillus) [inline src/main/java/twilightforest/biomes/RegistryBiomeEvent.java]: ?

**Comment by ghost** [comment]
@williewillus In RegistryBiomeEvent I aimed to have a MC 1.2.5 generator result for myself. Now, I reverted these changes.

File: src/main/java/twilightforest/block/ColorHandler.java [hunk 3]
```diff
@@ -263,9 +265,18 @@ public static void init() {
...
 	}
+	
+	private static int colorMultiplierFromItem( BlockColors blockColors, ItemStack stack, int tintIndex ) {
+		// NOTE: Handling possible forge bug. 
+		Item item = stack.getItem();
+		if( !(item instanceof ItemBlock) )
+			return -1;
+		
+		return blockColors.colorMultiplier(((ItemBlock)stack.getItem()).getBlock().getStateFromMeta(stack.getMetadata()), null, null, tintIndex);
+	}
```
Comment (williewillus) [inline src/main/java/twilightforest/block/ColorHandler.java]: ? is there an actual reproducible issue here? If so, report it to forge and put a comment to the forge issue here.

File: src/main/java/twilightforest/block/RegisterBlockEvent.java [hunk 4]
```diff
@@ -26,7 +26,11 @@ public static void onRegisterBlocks(RegistryEvent.Register<Block> event) {
 		blocks.register("hedge", (new BlockTFHedge()).setUnlocalizedName("TFHedge"));
 		blocks.register("boss_spawner", (new BlockTFBossSpawner()).setUnlocalizedName("TFBossSpawner"));
 		blocks.register("firefly_jar", (new BlockTFFireflyJar()).setUnlocalizedName("TFFireflyJar"));
-		blocks.register("twilight_plant", (new BlockTFPlant()).setUnlocalizedName("TFPlant"));
+//		blocks.register("twilight_plant", (new BlockTFPlant()).setUnlocalizedName("TFPlant"));
```
Comment (williewillus) [inline src/main/java/twilightforest/block/RegisterBlockEvent.java]: no commented things please :P

File: src/main/java/twilightforest/entity/EntityTFTowerGhast.java [hunk 5]
```diff
@@ -264,15 +264,15 @@ protected void updateAITasks() {
...
 
 	public int getAttackTimer() {
-		return dataManager.get(ATTACK_TIMER);
+		return (int)dataManager.get(ATTACK_TIMER);
 	}
 
 	public int getPrevAttackTimer() {
-		return dataManager.get(ATTACK_PREVTIMER);
+		return (int)dataManager.get(ATTACK_PREVTIMER);
 	}
```
Comment (williewillus) [inline src/main/java/twilightforest/entity/EntityTFTowerGhast.java]: these are unnecessary and unrelated to the PR

File: src/main/java/twilightforest/block/ColorHandler.java (hunk 3 above)
Comment (Tamaized) [reply src/main/java/twilightforest/block/ColorHandler.java]: Isn't this due to the Aurora double slab not having an itemblock, but this was already fixed in the 1.12 branch?

Comment (ghost) [reply src/main/java/twilightforest/block/ColorHandler.java]: It's reproducible in MC 1.11.2 Version, when picking up a lot of items at once. A random crash occurrs with a ClassCastException.

It's not known whether it is a forge bug or related to any coinstalled mod, like Optifine or s. e. . Sometimes invalid ItemStacks are passed containing items, having types which are not listed up in the argument list of a registerItemColorHandler call.

File: src/main/java/twilightforest/entity/EntityTFTowerGhast.java (hunk 5 above)
Comment (ghost) [reply src/main/java/twilightforest/entity/EntityTFTowerGhast.java]: Eclipse reported a compile error. Maybe it's an IDE version dependent issue. I use Eclipse Mars.

Feel free to remove these casts when merging.

**Comment by ghost** [comment]
Fixed the pull request.

Also I added an important comment at TFBiomeDecorator.decorate which should be considered when updating to MC 1.12.x .

**Comment by williewillus** [comment]
Please target 1.12, branch. Backports, in the unlikely event they do happen, happen after things in the main branch

**Comment by Drullkus** [comment]
Agreed, please target 1.12, thanks

**Comment by ghost** [comment]
Maybe there is a misunderstanding :) I'm not aiming for a backport. The thing is that I work on 1.11.2 and aim to restore a MC 1.2.5 look n' feel. I work towards a completely different direction. The sources I gave you are hopefully a useful contribution. But if you accept them, then they should be ported to the latest branch by TeamTwilight.

Of course, if I would contribute regularly then I would aim for the master branch ;)

**Comment by williewillus** [comment]
then this will not be taken unless it is submitted for 1.12, sorry

**Comment by ghost** [comment]
your choice.

**Pull request closed without merging**